neurox/data/writer.py

```diff
@@ -151,11 +153,17 @@ def open(self, num_layers):
         if self.decompose_layers:
             for layer_idx in self.layers:
                 local_filename = f"{self.filename[:-5]}-layer{layer_idx}.{self.filename[-4:]}"
-                _writer = self.base_writer(local_filename)
+                if self.base_writer == HDF5ActivationsWriter:
```

Comment:
I just took a quick look at how many decimal places can be accurately represented using a float16, and it seems like  anything smaller than 0.0005 cannot be precisely represented above a value of 1; so maybe dtype can play a role in the JSONWriter where we reduce the number of decimal places we save? I think currently currently we save 8 decimal places. What do you think?

Regardless, it might still be a good idea to pass in dtype and avoid this conditional here (more chances of issue if a new format is introduced in the future, or is someone or us decides to do something with dtype in the JSONWriter)

Reply:
This is what I found with a cursory search: https://stackoverflow.com/questions/872544/what-range-of-numbers-can-be-represented-in-a-16-32-and-64-bit-ieee-754-syste

If you can search a bit more and get a confident number that would be great (and maybe add it as a reference in the code itself)

Reply:
This is a nontrivial (but interesting!) issue. [This table](https://en.wikipedia.org/wiki/Half-precision_floating-point_format#Precision_limitations) shows the error intervals that depend on the absolute value of the half-precision float. With representations for a few sentences in BERT, I experienced a mean error of 6*10^{-7} between float16 and float32, with similarly low standard deviation of the errors. So that's quite precise. However, as you might know, some LMs have outlier neurons with abs(value)>10 (usually high negative activations). For these neurons, precision according to the table is $2^{-7} \approx 8*10^{-3}$ - much less precise than for values around 1. So the decimal places you save could really depend on the LM, your use case and the particular value of a neuron.

Possible solution: (i) Pass the dtype to the JSONActivationsWriter to avoid the if-statement, and (ii) save only 6 decimal places when dtype is float16? The latter change would roughly correspond to the mean error experienced for BERT, but it is a bit intransparent. So it could also be left out.

Should I put the link in one of the method docstrings of the transformers_extractor or writer? Or is a section in the docs needed that discusses which dtype to use? Some users might not think about the issue, and might want some guidance. Not sure where to put that though.

Reply:
I looked into the link you sent and also wrote a little script to see the differences (see end of message); as you say, we need approx 7 decimal digits to retain precision < 1 in float16. Doing some quick back of the envelope calculations:

For a bert model we have ~10k neurons from all layers
Lets say we have about 1000 sentences with 5 tokens each, so total 5k tokens

Number of activation values:
`5000 * 10000 = 50,000,000`

Using one less character for each of them (not strictly true because we can save < 8 decimal even now if the activation value does not have 8 decimal places, but for the sake of argument):
`50,000,000 bytes = ~50MB`

So 50MB saved for about 1000 sentences, which I'm not 100% is worth the intransparency you suggest. What do you think?

We should definitely pass the dtype to JSONWriter anyways even if we don't do anything with it.

```python
import numpy as np

np.set_printoptions(suppress=True)

# Check precisions
# Row 0: 0.xxxx
# Row 1: 1.xxxx
# Row 2: 10.xxxx
# Row 3: 100.xxxx
full_precision = np.zeros((4, 20), dtype="float32")
half_precision = np.zeros((4, 20), dtype="float16")

full_precision[1, :] = 1
half_precision[1, :] = 1

full_precision[2, :] = 10
half_precision[2, :] = 10

full_precision[3, :] = 100
half_precision[3, :] = 100

for i, e in enumerate(range(0, -20, -1)):
	full_precision[:, i] += 10 ** e
	half_precision[:, i] += 10 ** e

print(full_precision)
print(half_precision)
```

Reply:
I always work with hdf5, so I can't speak about the json usecase in practics. But I agree to your calculation: 50MB in your example is only a tiny fraction of the total file size for storing activations for 1000 sentences in json. So I agree, this small difference is probably not worth it. I'll pass the dtype parameter to the json writer but do nothing with it.